tests/base/hdf5_07.cc

```diff
@@ -0,0 +1,556 @@
+                                hyperslab_dimensions_c,
+                                hyperslab_offset_c,
+                                hyperslab_stride_c,
+                                hyperslab_count_c,
+                                hyperslab_block_c);
+      }
+    else
+      {
+        dataset.template write_none<Number>();
+      }
+
+    deallog << "Sum " + dataset_name << " " << container_name << "<"
+            << type_name << ">"
+            << " (Write): "
+            << container_sum(hyperslab_data_a) +
+                 container_sum(hyperslab_data_b) +
+                 container_sum(hyperslab_data_c)
+            << std::endl;
+    deallog << "Hyperslab_a sum " + dataset_name << " " << container_name << "<"
+            << type_name << ">"
+            << " (Write): " << container_sum(hyperslab_data_a) << std::endl;
+    deallog << "Hyperslab_b sum " + dataset_name << " " << container_name << "<"
+            << type_name << ">"
+            << " (Write): " << container_sum(hyperslab_data_b) << std::endl;
+    pcout << "IO mode " + dataset_name << " " << container_name << "<"
+          << type_name << ">"
+          << " (Write): " << dataset.get_io_mode() << std::endl;
+    pcout << "Local no collective cause " + dataset_name << " "
+          << container_name << "<" << type_name << ">"
+          << " (Write): " << dataset.get_local_no_collective_cause()
+          << std::endl;
+    pcout << "Global no collective cause " + dataset_name << " "
+          << container_name << "<" << type_name << ">"
+          << " (Write): " << dataset.get_global_no_collective_cause()
+          << std::endl;
+  }
+}
+
+
+
+int
+main(int argc, char **argv)
+{
+  // tests.h enables floating point exceptions in debug mode, but this test
+  // generates an (irrelevant) exception when run with more than one MPI
+  // process so disable them again:
+#if defined(DEBUG) && defined(DEAL_II_HAVE_FP_EXCEPTIONS)
+  {
+    const int current_fe_except = fegetexcept();
+    fedisableexcept(current_fe_except);
+  }
+#endif
+
+  initlog();
+
+  try
+    {
+      Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
+
+      MPI_Comm mpi_communicator = MPI_COMM_WORLD;
+
+      ConditionalOStream pcout(
+        std::cout, Utilities::MPI::this_mpi_process(mpi_communicator) == 0);
+
+      std::string                filename           = "test.h5";
+      const std::vector<hsize_t> dataset_dimensions = {50, 30};
+
+      HDF5::File data_file(filename,
+                           HDF5::File::FileAccessMode::create,
+                           mpi_communicator);
+
+      write_test<std::vector, double>(data_file,
+                                      dataset_dimensions,
+                                      mpi_communicator,
+                                      pcout);
+
+      AssertThrow(Utilities::MPI::this_mpi_process(mpi_communicator) == 0,
+                  ExcInternalError());
```

Comment:
If this is replaced by Assert, the program will hang as well.

Reply:
This code is equivalent to the following:
```
{
  const bool b = (Utilities::MPI::this_mpi_process(mpi_communicator) == 0);
  AssertThrow (b, ExcInternalError());
}
```
Can you find out whether it also hangs if you replace `AssertThrow` by `Assert` here? In other words, whether the problem is in the call to `Utilities::MPI::this_mpi_process` or in the `AssertThrow`?

Separately, do you actually need all of the code above? Or could `write_test()` be drastically reduced to just one piece?

Reply:
I don't have access to a computer this week.

I'll give you a thorough answer next week :)